toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -209,18 +291,142 @@ func resolveGraphNodes(dependencyGraph *pkggraph.PkgGraph, inputSummaryFile, out
+		if n.State == pkggraph.StateBuild {
+			err := downloadSingleDeltaRPM(realDependencyGraph, n, cloner, *outDir)
+			if err != nil {
+				return fmt.Errorf("failed to download delta RPM for build node %s: %w", n, err)
+			}
 		}
 	}
 
 	return
 }
 
+// downloadSingleDeltaRPM attempts to download a single delta RPM for a build node. If the delta RPM is available
+// it will be downloaded and the build node will be updated to point to the new RPM. The associated run node will
+// also be updated to point to the new RPM since the scheduler uses the run node to find the RPM to install. If a
+// delta RPM is not available, the build node will be left alone an no error will be returned.
+//   - realDependencyGraph: The graph to update
+//   - buildNode: The build node to update. This node should be from the real graph as we will be updating it directly.
+//     to find the actual build node in the graph.
+//   - cloner: The cloner to use to download the RPMs
+//   - deltaRpmDir: The directory to download the RPMs into
+func downloadSingleDeltaRPM(realDependencyGraph *pkggraph.PkgGraph, buildNode *pkggraph.PkgNode, cloner *rpmrepocloner.RpmRepoCloner, deltaRpmDir string) (err error) {
+	const downloadDependencies = false
+	var lookup *pkggraph.LookupNode
+
+	if buildNode.Type != pkggraph.TypeBuild {
+		err = fmt.Errorf("node '%s' is not a build node, can't download delta RPM", buildNode)
+		return err
+	}
+
+	// Replace all '/' with '_' in the package name to get a valid timestamp name
+	// e.g. "bin/ls" -> "bin_ls"
+	tsName := strings.ReplaceAll(buildNode.VersionedPkg.Name, "/", "_")
+	timestamp.StartEvent(fmt.Sprintf("downloading delta node %s", tsName), nil)
+	defer timestamp.StopEvent(nil)
+
+	lookup, err = realDependencyGraph.FindExactPkgNodeFromPkg(buildNode.VersionedPkg)
+	if err != nil {
+		err = fmt.Errorf("can't find build node '%s' in graph: %w", buildNode, err)
+		return err
+	}
+	if lookup == nil || lookup.RunNode == nil {
+		err = fmt.Errorf("can't find run lookup '%v' in graph", lookup)
+		return err
+	}
+
+	runNode := lookup.RunNode
+
+	// Get the final output path for the build node if we don't convert it to a delta node
+	originalRpmPath := buildNode.RpmPath
+	foundFinalRPM, err := file.PathExists(originalRpmPath)
+	if err != nil {
+		return fmt.Errorf("can't check if final RPM '%s' exists: %w", originalRpmPath, err)
+	}
+
+	// Only download dependencies for delta RPMs if we don't already have the RPM in the out/RPMS folder
+	if !foundFinalRPM {
+		// We have the expected rpm path, take the base name and strip .rpm off to get a name we can pass to tdnf
+		// to download the delta RPM
+		// e.g. "/home/user/repo/out/RPMS/x86_64/pkg-1.0-1.cm2.x86_64.rpm" -> "pkg-1.0-1.cm2.x86_64"
+		fullyQualifiedRpmName := filepath.Base(originalRpmPath)
+		fullyQualifiedRpmName = strings.TrimSuffix(fullyQualifiedRpmName, ".rpm")
+
+		cachedRPMPath := rpmPackageToRPMPath(fullyQualifiedRpmName, deltaRpmDir)
```

Comment:
Is this a correct path? At this point `fullyQualifiedRpmName` no longer has the `.rpm` suffix and I think the final file will have it.

Reply:
It should be good, `rpmPackageToRPMPath()` takes `<name>` and converts it to `/path/to/rpm_cache/cache/x86_64/<name>.rpm`. Agreed it's a bit silly to strip it and put it back but we needed the fuction for other stuff already